### PR TITLE
fix(binder): Incorrect cast when specifying columns

### LINKED ISSUE #7691: Incorrect cast when specifying columns
### Describe the bug

```SQL 
create table t (v1 real, v2 int, v3 varchar);
insert into t (v2, v1, v3) values (1, 2, 'a'), (3, 4, 'b');
```

Results in 

```
thread 'risingwave-batch-tasks' panicked at 'table source write chunk schema check failed: "column 0 should be Some(\"Float32\"), while stream chunk gives Some(\"Int32\")"', src/source/src/table.rs:102:14
```

I guess that we run into this, because we first call `cast_on_insert` ([here](https://github.com/risingwavelabs/risingwave/blob/20bdb72d5ff015d736fb582918cd551563139273/src/frontend/src/binder/insert.rs#L140) or in [bind_values here](https://github.com/risingwavelabs/risingwave/blob/20bdb72d5ff015d736fb582918cd551563139273/src/frontend/src/binder/insert.rs#L121)) and afterwords resolve where to insert the value ([here](https://github.com/risingwavelabs/risingwave/blob/20bdb72d5ff015d736fb582918cd551563139273/src/frontend/src/binder/insert.rs#L153)). 

I believe that we should also provide more tests which test this issue. 

### Additional context

I am currently on `main`, commit `2eea46208210d95c1975b7ea61f37674b6d2e832`

### PR DESCRIPTION
**Pull request by broccoliSpicy**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
fix #7691 

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [ ] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Comment by broccoliSpicy** [comment]
would you be a lamb and check out this pr for me @xxchan

File: src/frontend/src/binder/insert.rs
[the remark this answers is not in the document]
Comment (xxchan) [reply src/frontend/src/binder/insert.rs:1]: I will review it later. But if you can make it more reviewable by e.g., rewriting the commit history or explaining what's changed, I'm happy to review it earlier.

[the remark this answers is not in the document]
Comment (broccoliSpicy) [reply src/frontend/src/binder/insert.rs:1]: I was being inconsiderate...

File: src/frontend/src/binder/insert.rs [hunk 1]
```diff
@@ -102,14 +102,9 @@ impl Binder {
         let table_version_id = table_catalog.version_id().expect("table must be versioned");
         let columns_to_insert = table_catalog.columns_to_insert().cloned().collect_vec();
 
-        let expected_types: Vec<DataType> = columns_to_insert
```
Comment (broccoliSpicy) [inline src/frontend/src/binder/insert.rs:105]: the original expected_types is derived from the target table's column layout, which is a bug.
in queries like this 
```sql
create table t (v1 real, v2 int, v3 varchar);
insert into t (v2, v1, v3) values (1, 2, 'a'), (3, 4, 'b');
```
the expected types should follow the target column identifiers 

>  (v2, v1, v3)

File: src/frontend/src/binder/insert.rs [hunk 2]
```diff
@@ -135,151 +130,234 @@ impl Binder {
...
-                        with_ties: false,
-                        extra_order_exprs: vec![],
-                    },
-                    vec![],
-                    nulls_inserted,
-                )
+        let (returning_list, fields) = self.bind_returning_list(returning_items)?;
+        let is_returning = !returning_list.is_empty();
+
+        if !target_col_idents.is_empty() {
```
Comment (broccoliSpicy) [inline src/frontend/src/binder/insert.rs:136]: this line separate the queries into two cases
case 1:
```sql
insert into t (v2, v1, v3) values (1, 2, 'a'), (3, 4, 'b');
```
case 2:
```sql
insert into t values (1, 2, 'a'), (3, 4, 'b');
```

File: src/frontend/src/binder/insert.rs [hunk 3]
```diff
@@ -135,151 +130,234 @@ impl Binder {
...
+                    (bound, cast_exprs)
+                }
+            };
+            let insert = BoundInsert {
+                table_id,
+                table_version_id,
+                table_name,
+                owner,
+                row_id_index,
+                column_indices: (0..columns_to_insert.len()).collect::<Vec<usize>>(),
```
Comment (broccoliSpicy) [inline src/frontend/src/binder/insert.rs:349]: when user doesn't specify target column identifiers, the column indices should be ordinal

Comment (broccoliSpicy) [reply src/frontend/src/binder/insert.rs:349]: should I edit this into the code comments?

**Comment by broccoliSpicy** [comment]
actually this pr is quite minor, I am more than happy to explain the code as long as you don't think I am being garrulous, @xxchan,

File: src/frontend/src/binder/insert.rs [hunk 4]
```diff
@@ -135,151 +130,234 @@ impl Binder {
...
-            Query {
-                with: None,
-                body: SetExpr::Values(values),
-                order_by: order,
-                limit: None,
-                offset: None,
-                fetch: None,
-            } if order.is_empty() => {
-                let (values, nulls_inserted) =
-                    self.bind_values(values, Some(expected_types.clone()))?;
```
Comment (broccoliSpicy) [inline src/frontend/src/binder/insert.rs:170]: the original bind_value returns a tuple of (values, nulls_inserted), I don't think we need that nulls_inserted boolean, that is buggy logic

File: src/frontend/src/binder/insert.rs (hunk 2 above)
Comment (xxchan) [reply src/frontend/src/binder/insert.rs:136]: I feel the original big let block is more compact, but maybe it's also the cause of the bug. 🤔 Is it possible to reuse some code in the new version?

Comment (xxchan) [reply src/frontend/src/binder/insert.rs:136]: In other words, can we make it less spaghetti? 🙏🥹

File: src/frontend/src/binder/insert.rs [hunk 5]
```diff
@@ -135,151 +130,234 @@ impl Binder {
...
+                source,
+                cast_exprs,
+                returning_list,
+                returning_schema: if is_returning {
+                    Some(Schema { fields })
+                } else {
+                    None
+                },
+            };
+            Ok(insert)
```
[the remark this answers is not in the document]
Comment (broccoliSpicy) [reply src/frontend/src/binder/insert.rs]: long and repetitive indeed. I extracted the return logic, I could extract the second match arm into a helper function as well but seems like a fuss.
the second match is only a few lines of code
```rust
                query => {
                    let bound = self.bind_query(query)?;
                    let actual_types = bound.data_types();
                    let cast_exprs = match expected_types == actual_types {
                        true => vec![],
                        false => Self::cast_on_insert(
                            &expected_types,
                            actual_types
                                .into_iter()
                                .enumerate()
                                .map(|(i, t)| InputRef::new(i, t).into())
                                .collect(),
                        )?,
                    };
                    (bound, cast_exprs)
                }
 ```
I am not sure...

File: src/frontend/src/binder/insert.rs [hunk 6]
```diff
@@ -135,151 +130,234 @@ impl Binder {
...
+            // afterwards.
+            let (source, cast_exprs) = match source {
+                Query {
+                    with: None,
+                    body: SetExpr::Values(values),
+                    order_by: order,
+                    limit: None,
+                    offset: None,
+                    fetch: None,
+                } if order.is_empty() => {
```
[the remark this answers is not in the document]
Comment (xxchan) [reply src/frontend/src/binder/insert.rs]: Actually I mean putting the pattern matching into a function. 🤪 I've changed that. PTAL

File: src/frontend/src/binder/insert.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (xxchan) [reply src/frontend/src/binder/insert.rs]: The 2 second match arms being identical makes me think we should check it earlier outside the `if`.

[the remark this answers is not in the document]
Comment (xxchan) [reply src/frontend/src/binder/insert.rs]: Oh, the `expected_types` are different 😅

File: src/frontend/src/binder/insert.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (broccoliSpicy) [reply src/frontend/src/binder/insert.rs]: it's very kind of you

**Review by xxchan — COMMENTED** [review]
I found more code can be reused and made some refactoring during reviewing. PTAL if it looks good to you @broccoliSpicy

**Comment by broccoliSpicy** [comment]
> I found more code can be reused and made some refactoring during reviewing. PTAL if it looks good to you @broccoliSpicy

Your approach is way more concise, thank you :heart: @xxchan

**Review by xxchan — APPROVED** [review]

**Comment by xxchan** [comment]
Also thanks for your contribution ❤️ @broccoliSpicy

**Pull request merged**